File: wonyoung/algorithm_basic/3425.java
// 220414 3425 고스택
package wonyoung.algorithm_basic;

import java.io.*;
import java.util.*;

class Main {
    public static ArrayList<String[]> list = new ArrayList<>();
    public static Stack<long[]> stack = new Stack<>();
    public static long n;
    static long MAX = 1000000000;


    public static void main(String[] args) throws Exception {
        // for coding
        System.setIn(new FileInputStream("./input/input_3425.txt"));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st;

        while (true) {
            String cmd;
            while (true) {
                st = new StringTokenizer(br.readLine());

                if (!st.hasMoreTokens()) {
                    list.clear();
                    System.out.println();
                    continue;
                }
                cmd = st.nextToken();
                if (cmd.equals("NUM")) {
                    String num = st.nextToken();
                    list.add(new String[] { cmd, num });
                } else if (cmd.equals("END") || cmd.equals("QUIT")) {
                    break;
                } else {
                    list.add(new String[] { cmd });
                }
            }
            if ("QUIT".equals(cmd))
                break;


            boolean isNumber = false;
            st = new StringTokenizer(br.readLine());

            int tc = Integer.parseInt(st.nextToken());
            while (tc-- > 0) {
                st = new StringTokenizer(br.readLine());

                n = Long.parseLong(st.nextToken());
                isNumber = solution();
                if (!isNumber || stack.size() != 1) {
                    System.out.println("ERROR");
                } else {
                    if (Math.abs(stack.peek()[0]) > MAX)
                        System.out.println("ERROR");
                    else
                        System.out.println(stack.peek()[0]);
                }

                while (!stack.isEmpty())
                    stack.pop();
            }

            while (!list.isEmpty())
                list.remove(0);

        }

    }
    
// action for each command
    public static boolean solution() {
        stack.push(new long[] {n});
        for (String[] currList : list) {
            if (currList[0].equals("NUM")) {
                stack.push(new long[] {Long.parseLong(currList[1])});
            } else if (currList[0].equals("POP")) {
                if (stack.isEmpty())
                    return false;
                stack.pop();
            } else if (currList[0].equals("INV")) {
                if (stack.isEmpty())
                    return false;
                long a = stack.peek()[0];
                stack.pop();
                stack.push(new long[] {-a});
            } else if (currList[0].equals("DUP")) {
                if (stack.isEmpty())
                    return false;
                stack.push(stack.peek());
            } else if (currList[0].equals("SWP")) {
                if (stack.size() < 2)
                    return false;
                long num1 = stack.peek()[0];
                stack.pop();
                long num2 = stack.peek()[0];
                stack.pop();
                stack.push(new long[] {num1});
                stack.push(new long[] {num2});
            } else if (currList[0].equals("ADD")) {
                if (stack.size() < 2)
                    return false;
                long num1 = stack.peek()[0];
                stack.pop();
                long num2 = stack.peek()[0];
                stack.pop();
                stack.push(new long[] {num1+num2});
            } else if (currList[0].equals("SUB")) {
                if (stack.size() < 2)
                    return false;
                long num1 = stack.peek()[0];
                stack.pop();
                long num2 = stack.peek()[0];
                stack.pop();
                stack.push(new long[] {num2-num1});
            } else if (currList[0].equals("MUL")) {
                if (stack.size() < 2)
                    return false;
                long num1 = stack.peek()[0];
                stack.pop();
                long num2 = stack.peek()[0];
                stack.pop();
                stack.push(new long[] {num1*num2});
            } else if (currList[0].equals("DIV")) {
                if (stack.size() < 2)
                    return false;
                long num1 = stack.peek()[0];
                stack.pop();
                if (num1 == 0)
                    return false;
                long num2 = stack.peek()[0];
                stack.pop();
                stack.push(new long[] {num2/num1});
            } else if (currList[0].equals("MOD")) {
                if (stack.size() < 2)
                    return false;
                long num1 = stack.peek()[0];
                stack.pop();
                if (num1 == 0)
                    return false;
                long num2 = stack.peek()[0];
                stack.pop();
                stack.push(new long[] {num2%num1});
            }
        }
        return true;
    }
}

File: JOLLA/3425.py
#20220414 3425.py

import sys

input = sys.stdin.readline

def run(command, num):
    stack = [num]
    for cmd in command:
        if cmd[:3] == "NUM":
            n = int(cmd[4:])
            stack.append(n)
        elif not stack:
            return "ERROR"
        elif cmd == "POP":
            stack.pop()
        elif cmd == "INV":
            stack[-1] *= -1
        elif cmd == "DUP":
            stack.append(stack[-1])
        elif len(stack) ==1:
            return "ERROR"
        elif cmd == "SWP":
            stack[-1], stack[-2] = stack[-2], stack[-1]
        elif cmd == "ADD":
            temp = stack.pop() + stack.pop()
            if abs(temp) > 10 ** 9:
                return "ERROR"
            stack.append(temp)
        elif cmd == "SUB":
            temp = -stack.pop() + stack.pop()
            if abs(temp) > 10**9:
                return "ERROR"
            stack.append(temp)
        elif cmd == "MUL":
            temp = stack.pop() * stack.pop()
            if abs(temp) > 10**9:
                return "ERROR"
            stack.append(temp)
        elif cmd == "DIV":
            a, b = stack.pop(), stack.pop()
            if a == 0:
                return "ERROR"
            temp = abs(b) // abs(a)
            if (a>0 and b<0) or (a<0 and b>0):
                temp = -temp
            if abs(temp) > 10**9:
                return "ERROR"
            stack.append(temp)
        elif cmd == "MOD":
            a, b = stack.pop(), stack.pop()
            if a == 0:
                return "ERROR"
            temp = abs(b) % abs(a)
            if b < 0:
                temp = -temp
            if abs(temp) > 10**9:
                return "ERROR"
            stack.append(temp)
        else:
            return "ERROR"
    
    if len(stack) == 1:
        return stack[0]
    return "ERROR"


while True:
    command = []
    while True:
        cmd = input().strip()
        if cmd == "QUIT":
            quit()
        if cmd == "END":
            break
        command.append(cmd)

    n = int(input())

    for _ in range(n):
        num = int(input())
        print(run(command, num))
    ß
    print()
    input()



